[Pending Evolution] Super call enforcement

theblixguy opened the pull request:
Swift evolution thread: https://forums.swift.org/t/super-call-enforcement/38177
Swift evolution proposal: https://github.com/apple/swift-evolution/pull/1167

---

Introduce a new `@requiresSuper` attribute to control `super` calls in overridden methods. This is equivalent of `objc_requires_super` attribute in Objective-C. 

When a `@requiresSuper` attribute is present on a method, any overrides of that method should call the `super` method in their body. If the `super` call is missing, the compiler will emit an error. The error can be suppressed by inserting the `super` call and the compiler places no restrictions on where the call appears.

```swift
class A {
  @requiresSuper 
  func foo() { ... }
}

class B1: A {
  override func foo() { ... } // error: method override is missing 'super.foo()' call
}

class B2: A { 
  override func foo() { super.foo() } // okay
}
```

An optional message can be specified on the `@requiredSuper` attribute, which will be displayed with the error diagnostic:

```swift
class A {
  @requiresSuper("Call super as the final step in your implementation")
  func foo() { ... }
}

class B: A {
  override func foo() { ... } // error: method override is missing 'super.foo()' call: Call super as the final step in your implementation
}
```

A `@requiresSuper` attribute is also implicitly added for imported ObjC methods which have been annotated with `objc_requires_super`. This enables warnings (rather than errors, for better source compatibility) for any Objective-C method with such an attribute, such as `UIViewController.awakeFromNib()`.

```swift
import UIKit

class MyViewController1: UIViewController {
  override func awakeFromNib() {} // error: method override is missing 'super.awakeFromNib()' call
}

class MyViewController2: UIViewController {
  override func awakeFromNib() { super. awakeFromNib() } // okay
}
```

theblixguy committed:
[AST] Add 'requiresSuper' and 'ignoresSuper' attributes

theblixguy committed:
[AST] Add new diagnostics for 'requiresSuper' and 'ignoresSuper' attributes

theblixguy committed:
[Typechecker] Perform basic checks of 'requiresSuper' and 'ignoresSuper' attributes

theblixguy committed:
[Typechecker] Diagnose overridden method if it is missing 'ignoresSuper' attribute

theblixguy committed:
[Typechecker] Attach a fix-it to the diagnostic to remove the 'ignoresSuper' attribute if it's not valid

theblixguy committed:
[AST] Fix diagnostic wording for missing super call

theblixguy committed:
[MiscDiagnostics] Fix a bug that led to us incorrectly emitting the missing super call warning

theblixguy committed:
[ClangImporter] Import 'objc_requires_super' attribute as 'requiresSuper' attribute

theblixguy committed:
[Test] Add some basic tests

theblixguy committed:
[Parse] [AST] [Serialization] Support an optional message on the '@requiresSuper' attribute

theblixguy committed:
[Sema] Emit the optional message with missing super call error

theblixguy committed:
[Sema] Clone the '@requiresSuper' attribute to any overrides according to certain rules

theblixguy committed:
[Parse] Fix attribute range for '@requiresSuper'

theblixguy committed:
[ClangImporter] Fix a bug where the '@requiresSuper' attribute wasn't added to an imported method

theblixguy committed:
[Sema] Fix a bug where the merged '@requiresSuper' attribute had its implicit bit set to false

theblixguy committed:
[Test] Add some new test cases

theblixguy commented:
@swift-ci please smoke test

theblixguy commented:
@swift-ci please build toolchain

theblixguy commented:
Running source compatibility test suite to see if any projects break due to `objc_requires_super` being an error in Swift code.

@swift-ci please test source compatibility

theblixguy committed:
[Sema] Remove '@ignoresSuper' attribute and downgrade error to warning for imported ObjC decl

theblixguy commented:
@swift-ci please smoke test

theblixguy commented:
@swift-ci please test source compatibility

theblixguy commented:
@swift-ci please build toolchain

shahmishal commented:
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))

CodaFi commented:
We will need changes to `PrintAsObjC` to ensure that we emit this attribute in generated headers, and the requisite tests that all of that works.

CodaFi commented on include/swift/AST/Attr.def:
```diff
@@ -560,6 +560,11 @@ SIMPLE_DECL_ATTR(noDerivative, NoDerivative,
   100)
 
+DECL_ATTR(requiresSuper, RequiresSuper,
+  OnFunc |
```
`OnConstructor` as well, no?

theblixguy replied:
Good spot, I missed it. Fixed now!

CodaFi commented on include/swift/AST/Attr.def:
```diff
@@ -560,6 +560,11 @@ SIMPLE_DECL_ATTR(noDerivative, NoDerivative,
 
+DECL_ATTR(requiresSuper, RequiresSuper,
+  OnFunc |
+  ABIStableToAdd | ABIStableToRemove | APIStableToAdd | APIStableToRemove,
```
This is most certainly *not* API stable to add in a pure Swift context.

theblixguy replied:
You're right, I copy pasted it from the decl above but forgot to swap it with `APIBreakingToAdd` 🤦 Fixed though!

xedin commented on test/attr/requires_and_ignores_super.swift:
```diff
@@ -0,0 +1,90 @@
+
+class Invalid1A {
+  @requiresSuper
+  func foo(bar: Int) {}
```
I think we also need a more tests for error scenarios e.g. adding this attribute to a struct method, any (static or instance and/or final) property, any subscript, initializer, declaring it with an invalid message format etc.

CodaFi replied:
I'd like to see ObjC -> Swift tests for having this on `-init` and `-dealloc` as well.

theblixguy replied:
Added!

theblixguy commented:
Thanks for the feedback, I’ll address them shortly!

theblixguy committed:
[Misc] A couple of fixes that I forgot to do when rebasing

theblixguy committed:
[AST] Make 'requiresSuper' work with constructors and change 'APIStableToAdd' to 'APIBreakingToAdd

theblixguy committed:
[PrintAsObjC] Print 'NS_REQUIRES_SUPER' if method has 'requiresSuper' attribute

theblixguy committed:
[PrintAsObjC] Add 'NS_REQUIRES_SUPER' definition to the prologue header

theblixguy committed:
[AST] Support 'requiresSuper' on constructors as well

theblixguy committed:
[Sema] Run super call check on 'AbstractFunctionDecl' instead of just on 'FuncDecl'

theblixguy committed:
[Sema] Bail out early if 'requiresSuper' attr is invalid and also cast to 'AbstractFuncDecl'

theblixguy committed:
[Test] Add some more tests

theblixguy commented:
@CodaFi @xedin Apologies for the delay, been a little busy with life and took a break from contributions. I'm back now and I have addressed your feedback. Let me know if you have any further feedback. Thank you!

theblixguy committed:
[Serialization] Bump swiftmodule version number

theblixguy committed:
[AST] Tweak invalid requiresSuper message diagnostic slightly to add single quotes around attribute name

CodaFi commented on test/ClangImporter/requires_super_attribute.swift:
```diff
@@ -0,0 +1,21 @@
+// RUN: %target-swift-frontend -typecheck -import-objc-header %S/Inputs/objc-requires-super-obj.h %s -verify-ignore-unknown
+// REQUIRES: OS=macosx
```
I believe all this requires is objc_interop.

theblixguy replied:
Fixed!

theblixguy committed:
[Test] Use 'REQUIRES: objc_interop' and also fix a small mistake

CodaFi commented:
There's still an interesting question of semantics here if we're going to have an error. Super call enforcement in Objective-C is both a syntactic check and merely a warning because of it. If we're going to make this an error, I would expect the analysis to be properly semantic. That is, we would be checking super call enforcement along all edges of the CFG. As it stands, I can do this to silence the error (in both languages).

```
class Foo: NSObject {
  @requiresSuper init(chain: ()) {}
}

var condition = false

class Bar: Foo {
  override init() {
    if condition {
      super.init(chain: ())
    } else {
      super.init()
    }
  }
}
```

Is this intended?

theblixguy commented:
It’s not intended, no. Ideally this would be path sensitive but I initially implemented it in the type checker to make it work with the most common scenarios first and then iterate over it to catch these cases as well (potentially by moving the check into a SIL pass).

CodaFi commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -2651,6 +2651,17 @@ ERROR(override_method_different_generic_sig,none,
+ERROR(requires_super_attr_not_valid_final_method, none, 
+      "'@requiresSuper' cannot be applied to final method", ())
+ERROR(requires_super_attr_only_valid_on_class_method, none, 
+      "'@requiresSuper' can only be applied on non-final class methods", ())
```
Nit: `non-final class methods` means this to my mind

```
class Foo {
  @requiresSuper class func bar() {}
}
```

Perhaps 

```suggestion
      "'@requiresSuper' can only be applied on non-final methods on classes"
```

Or

```suggestion
      "'@requiresSuper' on method 'foo' of protocol 'Bar' is not allowed; only class methods may require calls to 'super'"
```

theblixguy replied:
This is a good suggestion! I can see why it would be confusing. I’ll make the amendment, thank you!

CodaFi commented:
My worry is that by implementing this as another syntactic check, we will have to go through another proposal process or wait for a new language mode to extend the check to be semantic - another cycle of breakage.

theblixguy commented:
Yeah, it’s definitely not _meant_ to be syntactic. I was just prototyping here, so implemented it in the type checker to validate the idea. I can move it into DI though for example. I think we already have some code that checks whether `self.init` was called on all paths and one for `yield` as well I believe. Perhaps I can reuse some existing logic to make it work here as well. Unless, you have a better idea in mind.

CodaFi commented:
Doing it as an extension of DI sounds good at first blush. @gottesmm can probably speak to that implementation strategy better though.

xedin commented:
> to make it work with the most common scenarios first and then iterate over it to catch these cases as well (potentially by moving the check into a SIL pass).

There is a major implication with approach like that. Every new invalid use turned into error means a source break for existing code which would be problematic, so if we want to make it an error I'd say we should do the outmost to catch as much as possible.

xedin commented:
@theblixguy I have talked to @gottesmm offline and he thinks that the best approach here would be to make it a diagnostic pass in SILOptimizer. You can take a look at how `YieldOnceCheck` is implemented which supposed to be very similar to what we want here. We'd need add a couple of SIL tests as well to make sure everything works as expected and have them use other expressions in the body besides `super.<method>` calls.

xedin reviewed (changes requested):
I'm just going to keep this as `Request changes` to keep that that I've looked at it.

gottesmm commented:
In general, we should avoid putting code into DI unless we absolutely have to. The pass is too big/complex. That is why I suggested that instead since there isn't a necessary reason to put this into DI, we should create a separate diagnostic pass. YieldOnceCheck does the same sort of stuff so it makes sense to just follow that.

alkozin commented:
Hello mates ✋
Is really somebody works on it?

I really like the idea of NS_REQUIRES_SUPER and can make this task done. 
But what we need to do for it?
Can we make the checklist?
